### PR TITLE
Add Qdrant Provider

### PR DESCRIPTION
**Pull request by Anush008**
This PR intends to add Qdrant as a supported provider by implementing an operator and hook that enables interfacing with Qdrant within Airflow DAGs.

[Qdrant](https://qdrant.tech/) is an open-source, Apache-licensed, vector search engine and database.

[[Discussion] Add Qdrant as a supported provider](https://lists.apache.org/thread/ojtvbs6bymvcjl25ojp1rnp24gqppf5z)

[[LAZY CONSENSUS] Adding an Airflow provider for Qdrant](https://lists.apache.org/thread/7h7gsgy2k8yx91l9l52yy1g93f1rjdqx)

### CODE REVIEW

File: airflow/providers/qdrant/hooks/qdrant.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        """Returns custom field behaviour."""
+        return {
+            "hidden_fields": ["schema", "login", "extra"],
+            "relabeling": {"password": "API Key"},
+        }
+
+    def __init__(self, conn_id: str = default_conn_name, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.conn_id = conn_id
+        self.get_conn()
```
Comment (hussein-awala) [inline airflow/providers/qdrant/hooks/qdrant.py]: Why do you call this method here?

**Review by hussein-awala — COMMENTED** [review]
I think that some of the hook tests are not needed since there is no wrapper for the client methods in the hook, you call them from the client and you check if they are called. Same for the hook integration tests. I don't have a strong opinion on this, let's wait for the other reviews.

File: airflow/providers/qdrant/hooks/qdrant.py (hunk 1 above)
Comment (Anush008) [reply airflow/providers/qdrant/hooks/qdrant.py]: Missed removing it. Great catch, thank you.
https://github.com/apache/airflow/pull/36805/commits/b97418434021ca335ae8ddf8c71f374c2a78a78a

**Comment by Anush008** [comment]
> I think that some of the hook tests are not needed since there is no wrapper for the client methods in the hook, you call them from the client and you check if they are called. Same for the hook integration tests. I don't have a strong opinion on this, let's wait for the other reviews.

Right. There were some wrapper methods before [this](https://github.com/apache/airflow/pull/36805/commits/bc880993c8afe7914dc17a84d06e5f8613f0cbae) commit. But then they seemed redundant as the caller can very well get the client through `get_conn()` and have access to all the methods. So I updated the mock hook tests accordingly.

File: airflow/providers/qdrant/provider.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+---
+package-name: apache-airflow-providers-qdrant
+
+name: Qdrant
+
+description: |
+    `Qdrant <https://qdrant.tech/documentation>`__
+
+state: not-ready
```
Comment (potiuk) [inline airflow/providers/qdrant/provider.yaml]: I think you shoudl change it to `ready` - we use not-ready only when we do not want to release provider **just yet** but this one seems quite, well , ready to be released.

Comment (Anush008) [reply airflow/providers/qdrant/provider.yaml]: https://github.com/apache/airflow/pull/36805/commits/cdf713ee0e71a05d8e263b0d0f04ce02657e44e9

File: airflow/providers/qdrant/provider.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+integrations:
+  - integration-name: Qdrant
+    external-doc-url: https://qdrant.tech/documentation
+    how-to-guide:
+      - /docs/apache-airflow-providers-qdrant/operators/qdrant.rst
+    tags: [software]
+
+dependencies:
+  - qdrant_client>=1.7.0
```
Comment (eladkal) [inline airflow/providers/qdrant/provider.yaml:40]:
```suggestion
  - qdrant_client>=1.7.0
  - apache-airflow>=2.6.0
```

File: docs/apache-airflow-providers-qdrant/connections.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    URL of the Qdrant instance to connect to. If specified, it overrides the ``host`` and ``port`` parameters.
+
+GRPC Port (optional)
+    GRPC port of the Qdrant instance to connect to. Defaults to ``6334``.
+
+Prefer GRPC (optional)
+    Whether to use GRPC for custom methods. Defaults to ``False``.
+
+HTTPS (optional)
+    Whether to use HTTPS for requests. Defaults to ``False``.
```
Comment (eladkal) [inline docs/apache-airflow-providers-qdrant/connections.rst]: Why default false?

Comment (Anush008) [reply docs/apache-airflow-providers-qdrant/connections.rst]: Since local Qdrant  deployments(at localhost) don't usually have TLS configured.
Most people start off with a local instance.

Comment (Anush008) [reply docs/apache-airflow-providers-qdrant/connections.rst]: Also, if an API key is set, `HTTPS` defaults to true. False otherwise.
`qdrant_client` handles this internally.

Edit: Just noticed, we need the value to be `None` for this inference.
Fixed in https://github.com/apache/airflow/pull/36805/commits/38ce62d3a5ab4a92a5d66b5f1561ab64383cffc6.

Comment (potiuk) [reply docs/apache-airflow-providers-qdrant/connections.rst]: Nice. Good catch @eladkal , thanks for insightful retrospection @Anush008 :)

Comment (potiuk) [reply docs/apache-airflow-providers-qdrant/connections.rst]: Would be nice to explain this inference in here.

Comment (Anush008) [reply docs/apache-airflow-providers-qdrant/connections.rst]: https://github.com/apache/airflow/pull/36805/commits/08d78845531a3aedd1eb91c8efa45e040dec3c1e

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Ah... Now ... after I refactored the contributing guide - there you need to solve conflicts  move things around I am afraid :( .. sorry for that @Anush008

File: airflow/providers/qdrant/hooks/qdrant.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    hook_name = "Qdrant"
+
+    @classmethod
+    def get_connection_form_widgets(cls) -> dict[str, Any]:
+        """Returns connection widgets to add to connection form."""
+        from flask_appbuilder.fieldwidgets import BS3TextFieldWidget
+        from flask_babel import lazy_gettext
+        from wtforms import BooleanField, IntegerField, StringField
+
+        return {
```
Comment (josh-fell) [inline airflow/providers/qdrant/hooks/qdrant.py:49]: Great use of `description`s in the form!

File: docs/apache-airflow-providers-qdrant/index.rst [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+Requirements
+------------
+
+The minimum Apache Airflow version supported by this provider package is ``2.5.0``.
+
+
+===================  ==================
+PIP package          Version required
+===================  ==================
+``apache-airflow``    ``>=2.5.0``
```
Comment (josh-fell) [inline docs/apache-airflow-providers-qdrant/index.rst]:
```suggestion
``apache-airflow``    ``>=2.6.0``
```

File: pyproject.toml [hunk 7]
```diff
@@ -533,6 +533,7 @@ alibaba = [
   "oss2>=2.14.0",
 ]
 amazon = [
+  "PyAthena>=3.0.10",
```
Comment (josh-fell) [inline pyproject.toml]: Is this a related change?

File: tests/providers/qdrant/operators/test_qdrant.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+from unittest.mock import patch
+
+import pytest
+
+from airflow.models import DAG
+from airflow.providers.qdrant.operators.qdrant import QdrantIngestOperator
+
+
+@pytest.fixture
+def dummy_dag():
```
Comment (josh-fell) [inline tests/providers/qdrant/operators/test_qdrant.py]: You could use the available [`dag_maker` fixture](https://github.com/apache/airflow/blob/8914e49551d8ae5ece7418950b011c1f338b4634/tests/conftest.py#L611) here if you'd like.

File: tests/system/providers/qdrant/example_dag_qdrant.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+from datetime import datetime
+
+from airflow import DAG
+from airflow.providers.qdrant.operators.qdrant import QdrantIngestOperator
+
+with DAG(
+    "example_qdrant_ingest",
+    schedule=None,
+    start_date=datetime(2024, 1, 1),
+    catchup=True,
```
Comment (josh-fell) [inline tests/system/providers/qdrant/example_dag_qdrant.py]:
```suggestion
    catchup=False,
```
Typically we make sure the examples have `catchup=False` just in case users (especially those new to Airflow) change the `schedule` or `start_date`. This way they aren't surprised by spawning DAG runs. Unless this is a deliberate decision (albeit the default)?

**Review by josh-fell — COMMENTED** [review]
This looks fantastic! Some small comments.

File: tests/providers/qdrant/operators/test_qdrant.py (hunk 8 above)
Comment (Anush008) [reply tests/providers/qdrant/operators/test_qdrant.py]: https://github.com/apache/airflow/pull/36805/commits/86c50b2e7f362e75b58a461ef5d3aa74e4ce8ae7

File: pyproject.toml (hunk 7 above)
Comment (Anush008) [reply pyproject.toml]: It came in with one of the merge conflict resolutions.

File: airflow/providers/qdrant/CHANGELOG.rst [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+``apache-airflow-providers-qdrant``
+
+Changelog
+---------
+
+1.0.0
+.....
+
+Initial version of the provider.
```
Comment (eladkal) [inline airflow/providers/qdrant/CHANGELOG.rst]:
```suggestion
* ``Initial version of the provider. (#36805)``
```

This will just help to populate this PR to the github issue status when we cut release

File: airflow/providers/qdrant/hooks/qdrant.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                "Example: `service/v1` will result in http://localhost:6333/service/v1/{qdrant-endpoint} for REST API.",
+            ),
+        }
+
+    @classmethod
+    def get_ui_field_behaviour(cls) -> dict[str, Any]:
+        """Returns custom field behaviour."""
+        return {
+            "hidden_fields": ["schema", "login", "extra"],
+            "relabeling": {"password": "API Key"},
```
Comment (eladkal) [inline airflow/providers/qdrant/hooks/qdrant.py:86]: just to clarify - this field is also masked when value is shown in the UI?

**Review by eladkal — APPROVED** [review]
LGTM

File: airflow/providers/qdrant/hooks/qdrant.py (hunk 11 above)
Comment (Anush008) [reply airflow/providers/qdrant/hooks/qdrant.py:86]: Yes, I think.

**Comment by potiuk** [comment]
FYI: @Anush008 : 

You will have to make your provider compatible with Airflow 2.7.0+  the compatibility checks detected, that your provider requires Pydantic 2 by, which is not supported in `Airflow 2.6.*`.  You will need to specify `apache-airflow>=2.6.0` in provider.yaml and exclude your provider for 2.6 checks in `BASE_PROVIDER_COMPATIBILITY_CHECKS`:

https://github.com/apache/airflow/blob/main/dev/breeze/src/airflow_breeze/global_constants.py#L467

**Comment by Anush008** [comment]
Hey.

> the compatibility checks detected, that your provider requires Pydantic 2 by, which is not supported in `Airflow 2.6.*`. 

`qdrant_client` uses Pydantic v1. The failing check compat seems to arise from Pydantic V2 use by a different provider.

**Comment by potiuk** [comment]
It's problem with Airflow -  2.6 has problem with Pydantic 1(that's why you need to limit it.  That's why you cannot install qdrant on it. As of 2.7 airflwo supports both an d in 2.8. only >2

**Comment by potiuk** [comment]
It's problem with Airflow -  2.6 has problem with Pydantic 1(that's why you need to limit it.  That's why you cannot install qdrant on it. As of 2.7 airflwo supports both an d in 2.8. only >2

Or so it looked like(I now am using a phone)

**Comment by potiuk** [comment]
FINALLY GREEN !!!.

**Review by potiuk — APPROVED** [review]
LGTM

**Comment by potiuk** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Comment by Anush008** [comment]
Thanks @potiuk, @eladkal for supporting this 🙏🙏.

Do let me know if there's anything else needed.
